### PR TITLE
12.0 mig coop shift

### PR DESCRIPTION
**Pull request by ajay-odoo**
Migrate Coop Shift to version 12

### CODE REVIEW

**Comment by ivantodorovich** [comment]
Can you fix LINT errors?

LINT:
```
coop_shift/wizard/update_shifts_wizard.py:26:1: E302 expected 2 blank lines, found 1
...
************* Module coop_shift.model.shift_template_registration
coop_shift/model/shift_template_registration.py:30: [R7980(consider-merging-classes-inherited), ShiftTemplateRegistration] Consider merging classes inherited to "event.registration" from model/shift_registration.py:44.
```

**Comment by ivantodorovich** [comment]
Build is failing because of missing `barcode_generator_partner`.

This is an OCA module, and it should be migrated from 11.0 to 12.0.
I'm creating the task in our project: https://www.druidoo.io/web#id=1478&action=367&model=project.task&view_type=form&menu_id=237
Is it migrated already? @krunal267 you remember something about it?



BUILD:
```
odoo.exceptions.UserError: ("You try to install module 'coop_shift' that depends on module 'barcodes_generator_partner'.\nBut the latter module is not available in your system.", '')
Message not found: 'Modules loaded.'
```

**Comment by ivantodorovich** [comment]
After fixing LINT errors, can you please squash commits into a single Migration to 12.0 commit?

image

**Comment by ivantodorovich** [comment]
@ajay-odoo Travis is fixed now, and the dependancy modules are available.
But the tests are raising errors. Can you please check?

**Comment by ivantodorovich** [comment]
@ajayjaviya can you please give priority to fixing this module?

it's a dependancy of coop_membership, and I'd like to see it merge asap

File: coop_shift/__init__.py
Comment (ivantodorovich) [inline coop_shift/__init__.py]: Please remove this long header from all file and replace by this style:

image


Don't forget to add 
```
# Copyright (C) 2020-Today: Druidoo (https://www.druidoo.io)
```

And make sure you keep the copyrights and authors present in the previous header block

File: coop_shift/__manifest__.py
Comment (ivantodorovich) [inline coop_shift/__manifest__.py]: Please check my comments about the `connector` dependancy on the spreadsheet file.
In 12.0, this module was separated into `connector` and `queue_job`.

I'm pretty sure we only need `queue_job` now, and we can replace this dependancy.

File: coop_shift/data/function.xml
Comment (ivantodorovich) [inline coop_shift/data/function.xml]: This empty file should be removed

File: coop_shift/data/ir_cron.xml
Comment (ivantodorovich) [inline coop_shift/data/ir_cron.xml]:
```suggestion
            <field name="name">Partner: Update Working State</field>
```

File: coop_shift/model/__init__.py
Comment (ivantodorovich) [inline coop_shift/model/__init__.py]:
```suggestion
```

File: coop_shift/model/__init__.py
Comment (ivantodorovich) [inline coop_shift/model/__init__.py]:
```suggestion
```

File: coop_shift/model/res_partner.py
Comment (ivantodorovich) [inline coop_shift/model/res_partner.py]: Please replace this big header with the new style, respecting original authors and copyrights, and adding druidoo

File: coop_shift/model/res_partner.py
Comment (ivantodorovich) [inline coop_shift/model/res_partner.py]: When you have long if statements, it's usually best to try something like this (it's more readable):

```py
if (
    extension.date_start <= fields.Date.today()
    and extension.date_stop > fields.Date.today()
):
    # do something
    # ...
```

File: coop_shift/model/shift_leave.py
Comment (ivantodorovich) [inline coop_shift/model/shift_leave.py]: Why was this removed?

I'm not sure of the functional implication of removing these lines. I'd prefer to keep them, just in case

File: coop_shift/model/shift_mail_registration.py
Comment (ivantodorovich) [inline coop_shift/model/shift_mail_registration.py]: Hmmm... Are we converting a date to string, and then back to date?
Why not just use `fields.datetime.now()` ? Am I missing something?

File: coop_shift/model/shift_mail_registration.py
Comment (ivantodorovich) [inline coop_shift/model/shift_mail_registration.py]: This `for rec in self` loop needs to be refactored.
You can't `return` inside the loop, it'll stop it.

If you call this function on an actual recordset with multiple records, it'll execute only the first one!


On way of refactoring this could be filtering the `self` recordset based on your conditions, and then use set operations to split the recordset and call the corresponding methods.

```py
records_to_execute = self.filtered(lambda rec: ...)
records_to_ignore = self - records_to_execute
records_to_skip = records_to_execute.filtered(lambda rec: rec.registration_id.partner_id.working_state in [..])
records_to_execute -= records_to_skip

if records_to_ignore:
    records_to_ignore.write({'mail_ignored': True})

return super(ShiftMailRegistration, records_to_execute).execute()
```

File: coop_shift/model/shift_registration.py
Comment (ivantodorovich) [inline coop_shift/model/shift_registration.py]: Please replace the ulgy header with the new style

File: coop_shift/model/shift_template.py
Comment (ivantodorovich) [inline coop_shift/model/shift_template.py]: Does this change have no impact ? Why remove this?

File: coop_shift/model/shift_template_ticket.py
Comment (ivantodorovich) [inline coop_shift/model/shift_template_ticket.py]: Remove old header style, replace with new

File: coop_shift/model/shift_type.py
Comment (ivantodorovich) [inline coop_shift/model/shift_type.py]: header

File: coop_shift/report/__init__.py
Comment (ivantodorovich) [inline coop_shift/report/__init__.py]: header

File: coop_shift/report/report_wallchart_common.py
Comment (ivantodorovich) [inline coop_shift/report/report_wallchart_common.py]: I see that you cleaned this report code. That's great. I have to say I'm not sure what you've done exactly, but we'll test it after its merged :)

Can you please remove the commented lines, though? There's no need to keep them, if they're no longer necessaries.
Are they, though? No longer necesarries

File: coop_shift/report/report_wallchart_ftop.py
Comment (ivantodorovich) [inline coop_shift/report/report_wallchart_ftop.py]: header

File: coop_shift/report/report_wallchart_template.py
Comment (ivantodorovich) [inline coop_shift/report/report_wallchart_template.py]: header

File: coop_shift/report/report_wallchart_template.py
Comment (ivantodorovich) [inline coop_shift/report/report_wallchart_template.py]: If no longer necessaries, remove

File: coop_shift/views/product_view.xml
Comment (ivantodorovich) [inline coop_shift/views/product_view.xml]: If no longer necessary, remove

File: coop_shift/wizard/add_template_registration.py
Comment (ivantodorovich) [inline coop_shift/wizard/add_template_registration.py]: header

File: coop_shift/wizard/__init__.py
Comment (ivantodorovich) [inline coop_shift/wizard/__init__.py]: header

File: coop_shift/wizard/create_shifts_wizard.py
Comment (ivantodorovich) [inline coop_shift/wizard/create_shifts_wizard.py]: header

**Review by ivantodorovich — CHANGES_REQUESTED** [review]
@ajayjaviya can you remove the `migrations` folder?

Also I've noticed there are many weird transformations on dates from/to strings.
In Odoo 12.0 date fields return date objects.

Can we avoid these transformations by refactoring the code a bit?

(don't squash the new changes when you push, so that it's easier for me to review what's changed)

File: coop_shift/model/__init__.py
Comment (ivantodorovich) [inline coop_shift/model/__init__.py:21]:
```suggestion
from . import ir_ui_menu

```

File: coop_shift/model/shift_shift.py
Comment (ivantodorovich) [inline coop_shift/model/shift_shift.py]: Can't this be achieved without using strings?

ie:

```py
self.date_begin = cur_date + timedelta(hours=self.shift_template_id.start_time)
```

File: coop_shift/model/shift_shift.py
Comment (ivantodorovich) [inline coop_shift/model/shift_shift.py]: Can't this be achieved without using strings?

File: coop_shift/model/shift_shift.py
Comment (ivantodorovich) [inline coop_shift/model/shift_shift.py]: What's this field for? `end_date_for_mail`
Looks like it's a formatted string used in templates, right?

In Odoo 12 we can just do this on templates, and it works as expected, using the user timezone/locale:

```html
<!-- For dates; -->
${format_date(object.date_end)}

<!-- or datetime -->
${format_datetime(object.date_end)}

<!-- This is also possible, and the result will be the same than previous code (datetime without seconds) -->
<!-- Although I wouldn't go for it because it ignores locale datetime format -->
${format_datetime(object.date_end, dt_format="dd/MM/YYYY HH:mm")}
```

Please see if you can remove it

File: coop_shift/model/shift_shift.py
Comment (ivantodorovich) [inline coop_shift/model/shift_shift.py]: Same for this field.
It shouldn't be necessary if we use `format_datetime` on the template.

**Review by ivantodorovich — CHANGES_REQUESTED** [review]

**Comment by ivantodorovich** [comment]
@ajayjaviya there are a few unresolved comments from my previous review, and some new ones.
Can you please check all of them?



Also travis is failing because of LINT:
```
coop_shift/model/shift_ticket.py:71:80: E501 line too long (82 > 79 characters)
coop_shift/model/shift_shift.py:343:80: E501 line too long (83 > 79 characters)
coop_shift/model/shift_template.py:341:80: E501 line too long (86 > 79 characters)
coop_shift/model/shift_template.py:604:80: E501 line too long (86 > 79 characters)
coop_shift/model/shift_template.py:620:80: E501 line too long (88 > 79 characters)
coop_shift/model/shift_template_registration_line.py:258:80: E501 line too long (86 > 79 characters)
```

File: coop_shift/model/shift_shift.py
Comment (ajay-odoo) [reply coop_shift/model/shift_shift.py]: @ivantodorovich : this field not used in email template at lest not in coop_shift module

File: coop_shift/model/shift_shift.py
Comment (ajay-odoo) [reply coop_shift/model/shift_shift.py]: @ivantodorovich : this field not used in email template at lest not in coop_shift module

Comment (ivantodorovich) [reply coop_shift/model/shift_shift.py]: Can you search for usage on all the repository? to see if it's used by other modules?
If it's not used, just remove it

**Comment by ivantodorovich** [comment]
Can you fix travis LINT errors?

Please try to do that autonomously every time :smile:

File: coop_shift/model/shift_leave.py
Comment (ajay-odoo) [reply coop_shift/model/shift_leave.py]: @ivantodorovich : this is just blocking duplicate leave but I don't see any functional use of it

**Comment by msdruidoo** [comment]
@ajay

We don't want to duplicate leaves

File: coop_shift/model/shift_leave.py
Comment (ivantodorovich) [reply coop_shift/model/shift_leave.py]: Can you please put it back?
We don't want to change any functional behaviour. Also Simon just confirmed this was intentional

**Review by ivantodorovich — CHANGES_REQUESTED** [review]
After you finish restoring the `copy` method restriction, can you squash your commits?
I believe this is ready to merge after that change

**Pull request merged**
